modules/globebrowsing/cache/memoryawaretilecache.cpp

```diff
@@ -32,51 +35,204 @@ namespace globebrowsing {
+    size_t sumTextureTypeSize = 0;
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        sumTextureTypeSize +=
+            it->second.first->tileTextureInitData().totalNumBytes();
+    }
+    size_t numTexturesPerType = estimatedSize / sumTextureTypeSize;
+    resetTextureContainerSize(numTexturesPerType);
+}
+
+void MemoryAwareTileCache::resetTextureContainerSize(size_t numTexturesPerTextureType) {
+    _numTextureBytesAllocatedOnCPU = 0;
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        it->second.first->reset(numTexturesPerTextureType);
+        it->second.second->clear();
+    }
 }
 
 bool MemoryAwareTileCache::exist(ProviderTileKey key) const {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    return _tileCache.exist(key);
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        if(it->second.second->exist(key)) {
+            return true;
+        }
+    }
+    return false;
 }
 
 Tile MemoryAwareTileCache::get(ProviderTileKey key) {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    return _tileCache.get(key);
+    for (TextureContainerMap::iterator it = _textureContainerMap.begin();
+        it != _textureContainerMap.end();
+        it++)
+    {
+        if(it->second.second->exist(key)) {
+            return it->second.second->get(key);
+        }
+    }
+    ghoul_assert(false, "");
 }
 
-void MemoryAwareTileCache::put(ProviderTileKey key, Tile tile) {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    _tileCache.put(key, tile);
+ghoul::opengl::Texture* MemoryAwareTileCache::getTexture(
+    const TileTextureInitData& initData)
+{
+    ghoul::opengl::Texture* texture;
+    // if this texture type does not exist among the texture containers
+    // it needs to be created
+    TileTextureInitData::HashKey initDataKey = initData.hashKey();
+    assureTextureContainerExists(initData);
+    // Now we know that the texture container exists,
+    // check if there are any unused textures
+    texture = _textureContainerMap[initDataKey].first->getTextureIfFree();
+    // Second option. No more textures available. Pop from the LRU cache
+    if (!texture) {
+        Tile oldTile = _textureContainerMap[initDataKey].second->popLRU();
+        // Use the old tile's texture
+        texture = oldTile.texture();
+    }
+    return texture;
 }
 
-void MemoryAwareTileCache::setMaximumSize(size_t maximumSize) {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-  _tileCache.setMaximumSize(maximumSize);
+void MemoryAwareTileCache::createTileAndPut(ProviderTileKey key,
+    std::shared_ptr<RawTile> rawTile)
+{
+    using ghoul::opengl::Texture;
```

Comment:
Just a different message, but it is internally based on ghoul_assert.  It's just a semantic difference: https://en.wikipedia.org/wiki/Design_by_contract